x/blockdb/recovery.go

```diff
@@ -0,0 +1,214 @@
+	var bh blockHeader
+	if totalDataSize-offset < uint64(sizeOfBlockHeader) {
+		return bh, fmt.Errorf("%w: not enough data for block header at offset %d", ErrCorrupted, offset)
+	}
+
+	dataFile, localOffset, err := s.getDataFileAndOffset(offset)
+	if err != nil {
+		return bh, fmt.Errorf("recovery: failed to get data file for offset %d: %w", offset, err)
+	}
+	bhBuf := make([]byte, sizeOfBlockHeader)
+	if _, err := dataFile.ReadAt(bhBuf, int64(localOffset)); err != nil {
+		return bh, fmt.Errorf("%w: error reading block header at offset %d: %w", ErrCorrupted, offset, err)
+	}
+	if err := bh.UnmarshalBinary(bhBuf); err != nil {
+		return bh, fmt.Errorf("%w: error deserializing block header at offset %d: %w", ErrCorrupted, offset, err)
+	}
+	if bh.Size == 0 || bh.Size > MaxBlockDataSize {
+		return bh, fmt.Errorf("%w: invalid block size in header at offset %d: %d", ErrCorrupted, offset, bh.Size)
+	}
+	if bh.Height < s.header.MinHeight || bh.Height == unsetHeight {
+		return bh, fmt.Errorf(
+			"%w: invalid block height in header at offset %d: found %d, expected >= %d",
+			ErrCorrupted, offset, bh.Height, s.header.MinHeight,
+		)
+	}
+	if bh.HeaderSize > bh.Size {
+		return bh, fmt.Errorf("%w: invalid block header size in header at offset %d: %d > %d", ErrCorrupted, offset, bh.HeaderSize, bh.Size)
+	}
+	expectedBlockEndOffset, err := safemath.Add(offset, uint64(sizeOfBlockHeader))
+	if err != nil {
+		return bh, fmt.Errorf("calculating block end offset would overflow at offset %d: %w", offset, err)
+	}
+	expectedBlockEndOffset, err = safemath.Add(expectedBlockEndOffset, uint64(bh.Size))
+	if err != nil {
+		return bh, fmt.Errorf("calculating block end offset would overflow at offset %d: %w", offset, err)
+	}
+	if expectedBlockEndOffset > totalDataSize {
+		return bh, fmt.Errorf("%w: block data out of bounds at offset %d", ErrCorrupted, offset)
+	}
+	blockData := make([]byte, bh.Size)
+	blockDataOffset, err := safemath.Add(localOffset, uint64(sizeOfBlockHeader))
+	if err != nil {
+		return bh, fmt.Errorf("calculating block data offset would overflow at offset %d: %w", offset, err)
+	}
+	if _, err := dataFile.ReadAt(blockData, int64(blockDataOffset)); err != nil {
+		return bh, fmt.Errorf("%w: failed to read block data at offset %d: %w", ErrCorrupted, offset, err)
+	}
+	calculatedChecksum := calculateChecksum(blockData)
+	if calculatedChecksum != bh.Checksum {
+		return bh, fmt.Errorf("%w: checksum mismatch for block at offset %d", ErrCorrupted, offset)
+	}
+
+	// Write index entry for this block
+	indexFileOffset, idxErr := s.indexEntryOffset(bh.Height)
+	if idxErr != nil {
+		return bh, fmt.Errorf("cannot get index offset for recovered block %d: %w", bh.Height, idxErr)
+	}
+	if err := s.writeIndexEntryAt(indexFileOffset, offset, bh.Size, bh.HeaderSize); err != nil {
+		return bh, fmt.Errorf("failed to update index for recovered block %d: %w", bh.Height, err)
+	}
+
+	if err := s.updateBlockHeights(bh.Height); err != nil {
```

Comment:
do we really need to do this each time for each block? Can't we do it after we scan all block entries?